### PR TITLE
New DerivedServices Architecture + Source-Cluster label

### PR DESCRIPTION
**Pull request by matthewgoodman13**
Affiliated with https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/47

*Description of changes:*
- Updated architecture to support multicluster services. Multiple derived services will be created for each cluster the multicluster service is exported from. See reference below.
    - Removed DerivedServiceAnnotation
- Added `multicluster.kubernetes.io/source-cluster` label to EndpointSlice with the value of the ClusterId in which it belongs to.
- Added unit tests for above functionality + edited existing tests to support above

*Testing:*
- Complete local testing - see PDF: [Testing for Pull Request #179.pdf](https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/files/9285061/Testing.for.Pull.Request.179.pdf)
- Added Unit Tests
- Kind Integration Suite runs without error

*Todo:*
- Add integration tests testing multicluster service discovery (Next PR!)

*Reference:*
CloudmapMCS

### CODE REVIEW

File: pkg/controllers/multicluster/cloudmap_controller.go [hunk 1]
```diff
@@ -184,23 +205,24 @@ func (r *CloudMapReconciler) createAndGetServiceImport(ctx context.Context, name
...
 
-	return r.getDerivedService(ctx, toCreate.Namespace, svcImport.Annotations[DerivedServiceAnnotation])
+	return r.getDerivedService(ctx, svcImport.Namespace, svcImport.Name, clusterId)
 }
 
-func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service) error {
+func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service, clusterId string) error {
 	existingSlicesList := discovery.EndpointSliceList{}
 	if err := r.Client.List(ctx, &existingSlicesList,
 		client.InNamespace(svc.Namespace), client.MatchingLabels{discovery.LabelServiceName: svc.Name}); err != nil {
```
Comment (runakash) [inline pkg/controllers/multicluster/cloudmap_controller.go:228]: Needs to also filter by label `multicluster.kubernetes.io/source-cluster`

File: pkg/controllers/multicluster/utils.go [hunk 2]
```diff
@@ -134,36 +137,72 @@ func PortsEqualIgnoreOrder(a, b []*model.Port) (equal bool) {
 	return true
 }
 
+func IPsEqualIgnoreOrder(a, b []string) (equal bool) {
```
Comment (runakash) [inline pkg/controllers/multicluster/utils.go:140]: Seems bit verbose. There is simpler solution mentioned here https://stackoverflow.com/a/67624073/791427

File: pkg/controllers/multicluster/cloudmap_controller.go [hunk 3]
```diff
@@ -131,41 +131,62 @@ func (r *CloudMapReconciler) reconcileService(ctx context.Context, svc *model.Se
...
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return err
+	// get or create derived Service for each cluster the service is a member of
+	derivedServices := make([]*v1.Service, 0, len(clusterIdToEndpointsMap))
+	for _, clusterId := range clusterIds {
+		endpoints := clusterIdToEndpointsMap[clusterId]
+		clusterImportedSvcPorts := ExtractServicePorts(endpoints)
+
+		derivedService, err := r.getDerivedService(ctx, svc.Namespace, svc.Name, clusterId)
```
Comment (runakash) [inline pkg/controllers/multicluster/cloudmap_controller.go:162]: We are not covering for the scenario when the service is removed from one of the cluster. In that case we should remove the derived service of that cluster. Best way to do this would be to fetch list of derived services, and delete those without any endpoints in cloudmap.

File: pkg/controllers/multicluster/utils.go [hunk 4]
```diff
@@ -177,7 +216,7 @@ func CreateDerivedServiceStruct(svcImport *multiclusterv1alpha1.ServiceImport, i
 	return &v1.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace:       svcImport.Namespace,
-			Name:            svcImport.Annotations[DerivedServiceAnnotation],
+			Name:            DerivedName(svcImport.Namespace, svcImport.Name, clusterId),
 			OwnerReferences: []metav1.OwnerReference{*ownerRef},
```
Comment (runakash) [inline pkg/controllers/multicluster/utils.go:220]: Please add `multicluster.kubernetes.io/source-cluster` label to derived service.

File: pkg/controllers/multicluster/cloudmap_controller.go [hunk 5]
```diff
@@ -245,11 +268,13 @@ func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport
...
-func (r *CloudMapReconciler) updateServiceImport(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, svc *v1.Service, importedSvcPorts []*model.Port) error {
+func (r *CloudMapReconciler) updateServiceImport(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, derivedServices []*v1.Service, importedSvcPorts []*model.Port) error {
 	updateRequired := false
-	if len(svcImport.Spec.IPs) != 1 || svcImport.Spec.IPs[0] != svc.Spec.ClusterIP {
-		r.Log.Debug("ServiceImport IP need update", "ServiceImport IPs", svcImport.Spec.IPs, "cluster IP", svc.Spec.ClusterIP)
-		svcImport.Spec.IPs = []string{svc.Spec.ClusterIP}
+
+	clusterIPs := GetClusterIpsFromServices(derivedServices)
+	if !IPsEqualIgnoreOrder(svcImport.Spec.IPs, clusterIPs) {
+		r.Log.Debug("ServiceImport IPs need update", "ServiceImport IPs", svcImport.Spec.IPs, "cluster IPs", clusterIPs)
```
Comment (runakash) [inline pkg/controllers/multicluster/cloudmap_controller.go:276]: Let's make this `Log.Info`

File: pkg/controllers/multicluster/utils.go [hunk 6]
```diff
@@ -134,36 +137,72 @@ func PortsEqualIgnoreOrder(a, b []*model.Port) (equal bool) {
...
+	clusterIPs := make([]string, 0)
+	for _, svc := range services {
+		clusterIPs = append(clusterIPs, svc.Spec.ClusterIP)
+	}
+	return clusterIPs
+}
+
 // DerivedName computes the "placeholder" name for an imported service
-func DerivedName(namespace string, name string) string {
+func DerivedName(namespace string, name string, clusterId string) string {
```
Comment (runakash) [inline pkg/controllers/multicluster/utils.go:168]: Can we write unit test for this? Make sure the output is deterministic.

**Review by runakash — CHANGES_REQUESTED** [review]

File: pkg/controllers/multicluster/utils.go (hunk 2 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/utils.go:140]: Can use one of those options. Is it worth bringing in new imports?

File: pkg/controllers/multicluster/utils.go (hunk 4 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/utils.go:220]: KEP does not specify this - but I do like the idea of the Service having reference to the source cluster it belongs to. Added.

File: pkg/controllers/multicluster/cloudmap_controller.go (hunk 1 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/cloudmap_controller.go:228]: Don't believe that is necessary. The svc.Name is the derivedService name. Therefore, filtering EndpointSlices based on the derived service name is sufficient to get its endpoint slices.

Derived Service Name will be unique as the name is built with hash(ns, svc-name, clusterId). If clusterId is different, then derived name will be unique and we do not need to filter by source cluster. 

(Unless we are worried about hash conflicts?)

File: pkg/controllers/multicluster/utils.go [hunk 7]
```diff
@@ -134,36 +137,72 @@ func PortsEqualIgnoreOrder(a, b []*model.Port) (equal bool) {
...
+		clusters = append(clusters, multiclusterv1alpha1.ClusterStatus{
+			Cluster: clusterId,
+		})
+	}
+
 	return &multiclusterv1alpha1.ServiceImport{
 		ObjectMeta: metav1.ObjectMeta{
-			Namespace:   namespace,
-			Name:        name,
-			Annotations: map[string]string{DerivedServiceAnnotation: DerivedName(namespace, name)},
```
Comment (runakash) [inline pkg/controllers/multicluster/utils.go:155]: Can we keep `DerivedServiceAnnotation` annotation? The reason being it will help with understanding the state if derived service is removed outside the controller. The value of annotation field is flexible, so we can put json here maybe?
Something like `[{ "cluster-id": "my-cluster1", "derived-service": "imported-xyz"}, { "cluster-id": "my-cluster2", "derived-service": "imported-pqr"}]`

File: pkg/controllers/multicluster/utils.go (hunk 2 above)
Comment (runakash) [reply pkg/controllers/multicluster/utils.go:140]: It's a standard google module, and quite small. I don't see concern. We possibly can use this for ports comparison too.

File: pkg/controllers/multicluster/utils.go (hunk 7 above)
Comment (runakash) [reply pkg/controllers/multicluster/utils.go:155]: @astaticvoid do you have an opinion on this?

Comment (astaticvoid) [reply pkg/controllers/multicluster/utils.go:155]: I agree we should keep the annotation, and the json representation makes sense to me.
Maybe we can be creative with where we create the derived name, possibly after the clusterIdToEndpointsMap.

File: pkg/controllers/multicluster/cloudmap_controller.go [hunk 8]
```diff
@@ -131,41 +131,62 @@ func (r *CloudMapReconciler) reconcileService(ctx context.Context, svc *model.Se
...
 	importedSvcPorts := ExtractServicePorts(svc.Endpoints)
 
+	clusterIdToEndpointsMap := make(map[string][]*model.Endpoint)
+	for _, ep := range svc.Endpoints {
+		clusterIdToEndpointsMap[ep.ClusterId] = append(clusterIdToEndpointsMap[ep.ClusterId], ep)
+	}
+
+	clusterIds := make([]string, 0, len(clusterIdToEndpointsMap))
+	for clusterId := range clusterIdToEndpointsMap {
+		clusterIds = append(clusterIds, clusterId)
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/cloudmap_controller.go:141]: Will keep as append. Underlying array structure does not change if length is correctly allocated. Also, this solution will be readable as compared to index approach.

File: pkg/controllers/multicluster/utils.go (hunk 7 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/utils.go:155]: Added back the annotation but will not be using it for anything.
Not worth using the annotation for retrieving the derived service name as it requires unmarshalling the json object every time we want to access and therefore is less efficient that just retrieving the name thru the DerivedName() hash.
Benchmarks:
```
c1 map[cluster-id:test-mcs-clusterid-2 derived-service:imported-us18hivtol]
unmarshalling time: 25.292µs
c2 imported-us18hivtol
taking DerivedService directly time: 6.367µs
```

File: integration/shared/scenarios/export_service.go [hunk 9]
```diff
@@ -60,8 +60,8 @@ func NewExportServiceScenario(cfg *aws.Config, nsName string, svcName string, po
 				Protocol:   string(v1.ProtocolTCP),
 			},
 			EndpointPort: endpointPort,
-			ClusterId:    test.ClusterId,
-			ClusterSetId: test.ClusterSetId,
+			ClusterId:    test.ClusterId1,
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply integration/shared/scenarios/export_service.go]: Updated the tests to pull from common.sh.

File: integration/janitor/runner/main.go [hunk 10]
```diff
@@ -9,12 +9,15 @@ import (
...
 		os.Exit(1)
 	}
 
-	j := janitor.NewDefaultJanitor()
 	nsName := os.Args[1]
+	clusterId := os.Args[2]
+	clusterSetId := os.Args[3]
+
+	j := janitor.NewDefaultJanitor(clusterId, clusterSetId)
 	j.Cleanup(context.TODO(), nsName)
```
Comment (astaticvoid) [inline integration/janitor/runner/main.go:22]: Not a  blocker but would prefer to use cluster set id in constructor, and cluster id in cleanup call, since this will presumably be used for cleaning up multiple clusters in a multi cluster integration test. We can postpone that change until that test is written however.

**Review by astaticvoid — APPROVED** [review]

File: integration/janitor/runner/main.go (hunk 10 above)
Comment (matthewgoodman13) [reply integration/janitor/runner/main.go:22]: Yes I do agree. Was planning to make some changes for how the janitor runs for multicluster tests

**Review by runakash — APPROVED** [review]

File: pkg/controllers/multicluster/cloudmap_controller.go [hunk 11]
```diff
@@ -180,24 +180,6 @@ func (r *CloudMapReconciler) reconcileService(ctx context.Context, svc *model.Se
 		if err = r.updateEndpointSlices(ctx, svcImport, endpoints, derivedService, clusterId); err != nil {
 			return err
 		}
-
-		derivedServices = append(derivedServices, derivedService)
-	}
-
-	// remove any existing derived services that do not have any endpoints in cloud map
-	existingDerivedServices := &v1.ServiceList{}
```
Comment (runakash) [inline pkg/controllers/multicluster/cloudmap_controller.go]: Is this code getting removed?

Comment (matthewgoodman13) [reply pkg/controllers/multicluster/cloudmap_controller.go]: should not be

Comment (matthewgoodman13) [reply pkg/controllers/multicluster/cloudmap_controller.go]: really weird - shows as added in my IDE - let me try again

**Review by astaticvoid — APPROVED** [review]
